File: 0031-next-permutation/0031-next-permutation.go
func nextPermutation(nums []int)  {
    var temp int
    if len(nums) <= 2{
        if len(nums) == 2{
            temp = nums[0]
            nums[0] = nums[1]
            nums[1] = temp
        }
    }else{
        index := -1
        for i:=len(nums)-2;i>=0;i--{
            if nums[i] < nums[i+1]{
                index = i
                break
            }
        }
        
        if index == -1{
            sort.Ints(nums)
        }else{
            for i:=len(nums)-1;i>index;i-- {
                if nums[index] < nums[i] {
                    
                    temp = nums[index]
                    nums[index] = nums[i]
                    nums[i] = temp
                    break
                }
            }
            
            index++
            end := len(nums)-1
            
            for index < end {
                temp = nums[index]
                nums[index] = nums[end]
                nums[end] = temp
                index++
                end--
            }            
            
        }
    }
}

File: 0031-next-permutation/0031-next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        if(nums.length > 2){
            int prefix=-1;
            for(int i=nums.length-2;i>=0;i--){
                if(nums[i] < nums[i+1]){
                    prefix = i;
                    break;
                }
            }
            if(prefix == -1){
                Arrays.sort(nums);
            }
            else{
                int temp;
                for(int i=nums.length-1;i>prefix;i--){
                    if(nums[prefix] < nums[i]){
                        temp = nums[prefix];
                        nums[prefix] = nums[i];
                        nums[i] = temp;
                        break;
                    }
                }
                prefix++;
                int end = nums.length-1;
                while(prefix < end){
                    temp = nums[prefix];
                    nums[prefix] = nums[end];
                    nums[end] = temp;
                    prefix++;
                    end--;
                }
            }
        }
        else{
            if(nums.length == 2){
                int temp;
                temp = nums[0];
                nums[0] = nums[1];
                nums[1] = temp;
            }
        }
    }
}
